Shopify/cli
# UI extensions serve in Node

### WHY are these changes introduced?
This PR is part of the effort of moving the logic for creating, building, and serving UI extensions from Node to Go.

### WHAT is this pull request doing?
This PR Duplicates the `serve` command from `packages/ui-extensions-go-cli` in `packages/app`.  Shopifolk are then opted into the new version unless they specify the `SHOPIFY_RUN_AS_USER=true` env var.

This code has a few concerns:

1. Creating a **bundler** to build extensions
2. Creating a **Payload Store** to maintain the state of all extensions
3. Creating a web server to respond to **fetch requests** to `/extensions/...`
4. Creating a **web socket server** at `/extensions/` that can be used to update the Payload Store, mediate messages, or inform clients of changes to the Payload Store
   
I'll explain these concerns in depth now:

#### Bundler
First, for each extension the bundler uses esbuild to watch for changes to the extension entry file.  E.g: `/extensions/[extension]/src/index.jsx`.  When ESBuild detects a change:

1. EsBuild builds the file to `/extensions/[extension]/dist/main.jsx`
2. EsBuild informs the Payload Store that of the build status.
3. The Payload Store then recalculates the data for that extension. 
4. The WebSockets then pass an update message to every client with the updated extension data

Second, for each extension the bundler uses [chokidar](https://www.npmjs.com/package/chokidar?activeTab=readme) to watch the `/extensions/[extension]/locales/` folder for changes to json files. When a file is added, edited or deleted:

1. Chokidar informs the PayloadStore that extension has changed
2. The Payload Store recalculates the data for that extension, including the localization data.

#### Payload Store
The Payload Store is responsible for keeping the state of extensions up to date.  At a high level it is an event emitter with methods to mutate it's internal state.  Whenever it updates it's internal state it emits events to allow listeners to do what they need to do.  In more detail it is responsible for:

1. Updating extension data: The websocket server could ask the Payload Store to update an extension.  An example of this would be when a user hovers, or hides an extension in the Dev Console.  Or the Bundler could ask to update the extension.
2. Updating app data:  Extension points can save arbitrary data to the app data that the Payload Store is responsible for.  In this situation an extension point would send a websocket message with new App Data, and the Payload Store would merge that data with existing data.
3. Informing Event Listeners on updates: The websocket server listens for when the Payload Store  updates data and informs clients of those updates.
4. Calculating data: One example of this is calculating the localization for any extension with localization files.  This involves parsing the JSON files on disk and converting those files into the Payload stores internal state.  Arguably this could be the responsibility of the bundler, but the current architecture is quite simple to reason about.

#### WebServer  for fetch Requests
The HTTP WebServer for fetch requests has the following end points:

1. `/extensions/dev-console` returns the UI and assets for the dev console 
2. `/extensions/[id]` returns the redirect URL to view the extension point for that extension.  That extension point will then connect to the server `extensions/[id]/assets/some/asset.js` to get the assets for that extension.  Right now this is just the `dist/main.js`
3. `/extensions/` returns the JSON data for all the extensions the developer has setup

In addition, the following applies to all endpoints

1. We allow Cross origin requests
2. We disable caching
3. We log requests for debugging

#### Websocket server
The web socket server is responsible for:

1. Sending the app & extensions state to clients when a client first creates a websocket connection.
2. Informing clients when there are changes to an extension.  An example of this is when a user edits a localization the end result is that the new extension data is sent via an `UPDATE` message to every client. Another example of this is when it responds to an `UPDATE` Message sent by a client.  It does this by asking the Payload Store to update the app or extension data with the data the client passed in the message. Once the Payload Store completes this operation it is the responsibility of the Websocket server to pass the new Payload Store data to all clients via an `UPDATE` message.
3. Mediating messages from clients.  Clients can send `DISPATCH` message.  The websocket server then passes this message on to every client.

### How to test your changes?

@dmhenry @pepicrft and @alfonso-noriega have tested each extension manually.  So I'll only add instructions for 2 extensions (Thanks Trish @vividviolet for these)

#### Checkout

1. Ensure your dev shop has the following beta force_checkout_one, checkout_one_argo_extensions
2. Run `yarn shopify app dev --path ./fixtures/app`.
3. Click on the extension preview link
4. Verify that you see an extension rendered inside the order summary
5. Make some changes to localization files in `/fixtures/app/extensions/checkout_ui`
6. Verify that the extension is reloaded and reflect the changes
7. Make some changes to the extension script in `/fixtures/app/extensions/checkout_ui`
8. Verify that the extension is reloaded and reflect the changes

#### Admin Product Subscription (published and unpublished extension)

1. Ensure your dev shop has the following beta argo_dev_portal, argo_dev_portal_mobile and the app is installed on your shop
2. Run `yarn shopify app dev --path ./fixtures/app`.
3. Click on the extension preview link
4. Verify that Web loads with the Dev Console appearing as an iframe at the bottom of the screen
5. Verify that you are redirect to a product details page
6. Navigate down to the Purchase Options card and click on Add purchase options
7. Verify that you can select your app and render the "Create new option" extension point
8. Make some changes to the extension script
9. Verify that the extension is reloaded and reflect the changes
10. Verify the following all work with the Product Subscription extension:
  - hovering on the Product Subscription extension should highlight that extension
  - clicking on the hide icon for the Product Subscription extension should make the extension unavailable. If you have 
  - multiple apps installed that powers the Subscription extension then clicking on "hide" should remove your test app from the list
  - clicking on the link for the Product Subscription extension should navigate to the product details page for the first product if you're not already on the page
  - clicking on the "refresh" button should force reload the extension
11. Testing a deployed the extension
  - Deploy your extension
  - Make some changes to the extension script
  - Click on the preview link again to load up the extension
  - Select your app and render the "Create new option" extension point
  - Make some changes to the extension script
  - Verify that the extension is reloaded and reflect the changes
  - Click the "hide" button on the Dev Console
  - Verify that you now see the published extension script - local changes should have no effect
  - Click the "show" button on the Dev Console
  - Verify that you see your local script rendered

### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
